interfaces/ros1/wavemap_ros/src/map_operations/publish_pointcloud_operation.cc

```diff
@@ -0,0 +1,132 @@
+#include "wavemap_ros/map_operations/publish_pointcloud_operation.h"
+
+#include <sensor_msgs/PointCloud.h>
+#include <sensor_msgs/PointCloud2.h>
+#include <sensor_msgs/point_cloud_conversion.h>
+#include <wavemap/core/indexing/index_conversions.h>
+#include <wavemap/core/map/hashed_chunked_wavelet_octree.h>
+#include <wavemap/core/map/hashed_wavelet_octree.h>
+#include <wavemap/core/utils/iterate/grid_iterator.h>
+#include <wavemap/core/utils/profiler_interface.h>
+#include <wavemap_ros_conversions/geometry_msg_conversions.h>
+
+namespace wavemap {
+DECLARE_CONFIG_MEMBERS(PublishPointcloudOperationConfig,
+                      (once_every)
+                      (occupancy_threshold_log_odds)
+                      (only_publish_changed_blocks)
+                      (topic));
+
+bool PublishPointcloudOperationConfig::isValid(bool verbose) const {
+  bool all_valid = true;
+
+  all_valid &= IS_PARAM_GT(once_every, 0.f, verbose);
+  all_valid &= IS_PARAM_NE(topic, std::string(), verbose);
+
+  return all_valid;
+}
+
+PublishPointcloudOperation::PublishPointcloudOperation(
+    const PublishPointcloudOperationConfig& config, MapBase::Ptr occupancy_map,
+    std::string world_frame, ros::NodeHandle nh_private)
+    : MapOperationBase(std::move(occupancy_map)),
+      config_(config.checkValid()),
+      world_frame_(std::move(world_frame)) {
+  pointcloud_pub_ =
+      nh_private.advertise<sensor_msgs::PointCloud2>(config_.topic, 10);
+}
+
+bool PublishPointcloudOperation::shouldRun(const ros::Time& current_time) {
+  return config_.once_every < (current_time - last_run_timestamp_).toSec();
+}
+
+void PublishPointcloudOperation::run(bool force_run) {
+  const ros::Time current_time = ros::Time::now();
+  if (force_run || shouldRun(current_time)) {
+    publishPointcloud(current_time);
+    last_run_timestamp_ = current_time;
+  }
+}
+
+void PublishPointcloudOperation::publishPointcloud(
+    const ros::Time& current_time) {
+  ProfilerZoneScoped;
+  // If the map is empty, there's no work to do
+  if (occupancy_map_->empty()) {
+    return;
+  }
+
+  // Create the message and set the header
+  sensor_msgs::PointCloud obstacle_pointcloud_msg;
+  obstacle_pointcloud_msg.header.frame_id = world_frame_;
+  obstacle_pointcloud_msg.header.stamp = current_time;
+
+  // Define a functor that converts map leaf nodes into pointcloud points
+  auto add_points_for_leaf_node =
```

Comment:
This creates non-uniform density pointclouds for the non highest resolution nodes. Would it be sensible to create a point for each highest possible resolution voxel as opposed to the center of the larger voxel?

Reply:
The code already does this :) If the node is not at the highest resolution (node_index.height != 0), we iterate over the grid it covers at the highest resolution to add uniform density points [on lines 76 to 86](https://github.com/ethz-asl/wavemap/blob/develop/v2.0/interfaces/ros1/wavemap_ros/src/map_operations/publish_pointcloud_operation.cc#L76-L86).